### PR TITLE
Book 3 review, chapters 4-6

### PR DESCRIPTION
**Pull request by trevordblack**

### CODE REVIEW

**Comment by trevordblack** [comment]
When I said that most of the edits to book 3 should be constrained to chapters 1-3, that was hopelessly, hopelessly naive.

File: books/RayTracingTheRestOfYourLife.html [hunk 1]
```diff
@@ -1530,7 +1543,7 @@
 Playing with Importance Sampling
 ====================================================================================================
 <div class='together'>
-Our goal over the next two chapters is to instrument our program to send a bunch of extra rays
+Our goal over the next four chapters is to instrument our program to send a bunch of extra rays
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:1546]: Perhaps “Our goal over the following chapters is to …”?

File: books/RayTracingTheRestOfYourLife.html [hunk 2]
```diff
@@ -1680,67 +1679,81 @@
...
+            if (!rec.mat->scatter(r, rec, attenuation, scattered))
                 return emitted;
 
-            return emitted
-                 + albedo * rec.mat->scattering_pdf(r, rec, scattered)
-                          * ray_color(scattered, depth-1) / pdf;
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++ highlight
+            double scattering_pdf = rec.mat->scattering_pdf(r, rec, scattered);
+            double pdf = scattering_pdf;
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:1692]: I don't think this is right. If it was, then `scattering_pdf` and `pdf` below would just cancel out.

Ah, never mind. You're doing this for exposition.

File: books/RayTracingTheRestOfYourLife.html [hunk 3]
```diff
@@ -1680,67 +1679,81 @@
...
+            return emitted + scattered;
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
         }
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     [Listing [ray-color-impsample]: <kbd>[scene.h]</kbd>
     The ray_color function, modified for importance sampling]
 </div>
 
-You should get exactly the same picture.
+You should get exactly the same picture. Which, _should make sense_, as the scattered part of
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:1705]: “Which _should make sense_, as the …” (strike comma after Which).

File: books/RayTracingTheRestOfYourLife.html [hunk 4]
```diff
@@ -1750,73 +1763,93 @@
...
-technique, but we are no longer treating our objects as Lambertian. They are a type of diffuse
-material, but Lambertian requires a $\cos(\theta_o)$ sampling pattern. So any differences that we
-see is because we are using a slightly different material. We're going to uniformly scatter random
-directions from the hemisphere and perfectly match the PDF, given by
-$\mathit{pScatter}(\omega_o) = p(\omega_o) = \frac{1}{2\pi}$.
+To confirm our understanding, let's try a different scattering distribution. For this one, we'll
+attempt to repeat the uniform hemispherical scattering from the first book. There's nothing wrong
+with this technique, but we are no longer treating our objects as Lambertian. Lambertian is a
+specific type of diffuse material that requires a $\cos(\theta_o)$ scattering distribution.
+Uniform hemispherical scattering is another diffuse material, albeit a different one. If we keep the
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:1791]: “… another diffuse material ~, albeit a different one~.”

File: books/RayTracingTheRestOfYourLife.html [hunk 5]
```diff
@@ -1750,73 +1763,93 @@
...
+material the same but change the PDF, as we did in last section, we will still converge on the same
+answer, but our convergence may take more or less samples. However, if we change the material, we
+will have fundamentally changed the render and the algorithm will converge on a different answer. So
+when we replace Lambertian diffuse with uniform hemispherical diffuse we should expect that the
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:1795]: “we should expect ~that~ the outcome of our render …”

**Review by hollasch — APPROVED** [review]

**Comment by trevordblack** [comment]
`color scattered` needs to be renamed. as `ray scattered` exists within the same scope

File: books/RayTracingTheRestOfYourLife.html (hunk 1 above)
Comment (hollasch) [reply books/RayTracingTheRestOfYourLife.html:1546]: “over the next several”?

File: books/RayTracingTheRestOfYourLife.html [hunk 6]
```diff
@@ -1425,32 +1426,34 @@
...
-mirror-like as your viewing angle (incident angle) approaches grazing. Lastly, the scattering PDF
-can also depend on the scattering position: $\mathit{pScatter}(\mathbf{x}, \omega_i, \omega_o)$. The
+scattering PDF will vary with outgoing direction: $\mathit{pScatter}(\omega_o)$. The scattering PDF
+can also vary with _incident direction_: $\mathit{pScatter}(\omega_i, \omega_o)$. You can see this
+varying with incident direction when you look at reflections off a road -- they become mirror-like
+as your viewing angle (incident angle) approaches grazing. The scattering PDF can vary with the
+wavelength of the light: $\mathit{pScatter}(\omega_i, \omega_o, \lambda)$. A good example
+of this is a prism refracting white light into a rainbow. Lastly, the scattering PDF can also depend
+on the scattering position: $\mathit{pScatter}(\mathbf{x}, \omega_i, \omega_o, \lambda)$. The
 $\mathbf{x}$ is just a vector representing the scattering position: $\mathbf{x} = (x, y, z)$. The
```
Comment (TheRayTracer) [inline books/RayTracingTheRestOfYourLife.html:1440]: Just a vector
or
Just a point
?

File: books/RayTracingTheRestOfYourLife.html [hunk 7]
```diff
@@ -1680,67 +1679,81 @@
...
+            color attenuation;
+            color emitted = rec.mat->emitted(rec.u, rec.v, rec.p);
+
+            if (!rec.mat->scatter(r, rec, attenuation, scattered))
+                return emitted;
+
+            double scattering_pdf = rec.mat->scattering_pdf(r, rec, scattered);
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++ highlight
-            return 0.5 / pi;
+            double pdf = 1 / (2*pi);
```
Comment (TheRayTracer) [inline books/RayTracingTheRestOfYourLife.html:1745]: I think a note should be made somewhere that the pdf and scattering pdf are currently cancelling each other out. By multiplying by `0.5 / pi`, and then dividing by `0.5 / pi` in the same expression.

image

**Comment by trevordblack** [comment]
Fixes #933

**Comment by trevordblack** [comment]
Need to address #975

File: books/RayTracingTheRestOfYourLife.html (hunk 7 above)
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:1745]: ?

I think you're looking over the book before this CL.
I explicitly call out that they cancel each other in the latest change

Comment (TheRayTracer) [reply books/RayTracingTheRestOfYourLife.html:1745]: Awesome. It's great that you've called this observation out. to make it clear

Comment (TheRayTracer) [reply books/RayTracingTheRestOfYourLife.html:1745]: Feel free to make as Resolved (I don't seem to have that option)

File: books/RayTracingTheRestOfYourLife.html (hunk 6 above)
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:1440]: iiiiiiiiiiii

it's both. Ugh.
It's a mathematical vector, which is subtly different from the vector we've been using

Comment (TheRayTracer) [reply books/RayTracingTheRestOfYourLife.html:1440]: Your choice. I just thought it was worth the callout.

File: books/RayTracingTheRestOfYourLife.html (hunk 7 above)
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:1744]: Cool. Thanks for the feedback!

**Pull request merged**